File: server/utils/publisherKey.js
const { cleanIngestSourceLabel, publisherNameFromPost } = require('./serializers');

/** Stable key for follow/filter (ASCII slug + preserves Indic via unicode). */
function publisherKeyFromName(raw) {
  const cleaned = cleanIngestSourceLabel(raw) || String(raw || '').trim();
  if (!cleaned) return '';
  return cleaned
    .toLowerCase()
    .replace(/[^a-z0-9\u0900-\u097F\u0C00-\u0C7F]+/g, '-')
    .replace(/^-+|-+$/g, '')
    .slice(0, 120);
}

function publisherIdentityFromPost(post) {
  const name = publisherNameFromPost(post);
  const key = publisherKeyFromName(name);
  return { publisherKey: key, publisherName: name };
}

module.exports = {
  publisherKeyFromName,
  publisherNameFromPost,
  publisherIdentityFromPost,
};

File: flutter_app/lib/utils/publisher_key.dart
import '../models/models.dart';

/// Stable publisher key — keep in sync with server `publisherKey.js`.
String publisherKeyFromName(String raw) {
  final cleaned = cleanIngestSourceLabel(raw) ?? raw.trim();
  if (cleaned.isEmpty) return '';
  var key = cleaned
      .toLowerCase()
      .replaceAll(RegExp(r'[^a-z0-9\u0900-\u097F\u0C00-\u0C7F]+'), '-')
      .replaceAll(RegExp(r'^-+|-+$'), '');
  if (key.length > 120) key = key.substring(0, 120);
  return key;
}

String publisherKeyForPost(NewsPost post) =>
    publisherKeyFromName(post.displaySourceName);
